connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/connect/client/GrpcRetryHandler.scala

```diff
@@ -0,0 +1,187 @@
+
+  /**
+   * Generalizes the retry logic for RPC calls that return an iterator.
+   * @param request
+   *   The request to send to the server.
+   * @param call
+   *   The function that calls the RPC.
+   * @tparam T
+   *   The type of the request.
+   * @tparam U
+   *   The type of the response.
+   */
+  class RetryIterator[T, U](request: T, call: T => java.util.Iterator[U])
+      extends java.util.Iterator[U] {
+
+    private var opened = false // we only retry if it fails on first call when using the iterator
+    private var iterator = call(request)
+
+    private def retryIter[V](f: java.util.Iterator[U] => V) = {
+      if (!opened) {
+        opened = true
+        var firstTry = true
+        retry {
+          if (firstTry) {
+            // on first try, we use the initial iterator.
+            firstTry = false
+          } else {
+            // on retry, we need to call the RPC again.
+            iterator = call(request)
+          }
+          f(iterator)
+        }
+      } else {
+        f(iterator)
+      }
+    }
+
+    override def next: U = {
+      retryIter(_.next())
+    }
+
+    override def hasNext: Boolean = {
+      retryIter(_.hasNext())
+    }
+  }
+
+  object RetryIterator {
+    def apply[T, U](request: T, call: T => java.util.Iterator[U]): RetryIterator[T, U] =
+      new RetryIterator(request, call)
+  }
+
+  /**
+   * Generalizes the retry logic for RPC calls that return a StreamObserver.
+   * @param request
+   *   The request to send to the server.
+   * @param call
+   *   The function that calls the RPC.
+   * @tparam T
+   *   The type of the request.
+   * @tparam U
+   *   The type of the response.
+   */
+  class RetryStreamObserver[T, U](request: T, call: T => StreamObserver[U])
+      extends StreamObserver[U] {
+
+    private var firstOnNext = false // only retries on first onNext call
+    private var streamObserver = call(request)
+    override def onNext(v: U): Unit = {
+      if (!firstOnNext) {
+        firstOnNext = false
+        var firstTry = true
+        retry {
+          if (firstTry) {
+            // on first try, we use the initial streamObserver.
+            firstTry = false
+          } else {
+            // on retry, we need to call the RPC again.
+            streamObserver = call(request)
+          }
+          streamObserver.onNext(v)
+        }
+      } else {
+        streamObserver.onNext(v)
+      }
+    }
+    override def onError(throwable: Throwable): Unit = {
+      firstOnNext = false
+      streamObserver.onError(throwable)
+    }
+    override def onCompleted(): Unit = {
+      firstOnNext = false
+      streamObserver.onCompleted()
+    }
+  }
+
+  object RetryStreamObserver {
+    def apply[T, U](request: T, call: T => StreamObserver[U]): RetryStreamObserver[T, U] =
+      new RetryStreamObserver(request, call)
+  }
+}
+
+private[client] object GrpcRetryHandler {
+  def apply(retryPolicy: RetryPolicy): GrpcRetryHandler = new GrpcRetryHandler(retryPolicy)
```

Comment:
do we need this apply? I think it would be fine for the users to use the normal constructor.